go/types/path.go

```diff
@@ -91,6 +94,15 @@ func constructPath(p Path, str string) (Path, error) {
 		p = append(p, part)
 		return constructPath(p, rem)
 
+	case '@':
+		ann, rem := getAnnotation(tail)
+		switch ann {
+		case "type":
+			return constructPath(append(p, TypePart{}), rem)
```

Comment:
```go
// (Implemented by IndexPath and HashIndexPath)
type keyIndexable interface {
  setIntoKey(intoKey bool)
}
...
case "key":
  if len(p) == 0 {
    return Path{}, fmt.Errorf("Cannot @key index at the start of a path")
  }
  lastPart := p[len(p) - 1]
  if ki, ok := lastPart.(keyIndexable); ok {
    ki.setIntoKey(true)
  } else {
    return Path{}, fmt.Errorf("Cannot @key index into %s", lastPart.String())
  }
```
With appropriate other changes of course (e.g. removing the `NewHashIndexIntoKeyPath`-style methods). This would at least be equivalent with the existing code, and address the TODO.